src/rdb.c

```diff
@@ -1410,11 +1410,24 @@ int rdbSaveBackground(char *filename, rdbSaveInfo *rsi) {
     return C_OK; /* unreached */
 }
 
-void rdbRemoveTempFile(pid_t childpid) {
+/* Note that we may call this function in signal handle 'sigShutdownHandler',
+ * so we need guarantee all functions we call are async-signal-safe.
+ * If 'force' is set to 1, we'll remote file directly, since 'bg_unlink' is
+ * not async-signal-safe, 'force' must be set to 1 in signal handle. */
+void rdbRemoveTempFile(pid_t childpid, int force) {
     char tmpfile[256];
+    char pid[32];
 
-    snprintf(tmpfile,sizeof(tmpfile),"temp-%d.rdb", (int) childpid);
-    unlink(tmpfile);
+    /* Generate temp rdb file name using aync-signal safe functions. */
+    int pid_len = ll2string(pid, sizeof(pid), childpid);
+    strcpy(tmpfile, "temp-");
+    strncpy(tmpfile+5, pid, pid_len);
+    strcpy(tmpfile+5+pid_len, ".rdb");
+
+    if (force)
+        unlink(tmpfile);
```

Comment:
how about renaming `force` to `from_signal`?
and how about using the following:
```c
if (from_signal) {
    /* bg_unlink is not async-signal-safe, but in this case we don't really
     * need to close the fd, it'll be released when the process exists. */
    int fd = open(tmpfile);
    UNUSED(fd);
    unlink(tmpfile);
} else {
    bg_unlink(tmpfile);
}
```

Reply:
For rename `force` to `from_signal`, I think it is OK.

But we leave a non-closed fd deliberately,  release it depend on OS, I  truly don't think it is a good idea.  What's more, if users want to quick ASAP, the second 'kill -SIGINT $pid'  is short from the first, so temp rdb will not be very big, so we needn't do that.

Reply:
don't we already have the situation of exiting with a pending bg_unlink? (after removing the loop you wanted to add in prepareForShutdown)? if we do have that situation in one place i don't see why we don't want it in another.

also, if the OS doesn't clean a stale inode from the disk in that case, isn't that a serious (possibly rare) bug in the OS?

i didn't understand your argument about the second kill and file not being very big.

i don't object to merge the current version of the PR (maybe with the small rename i added). but i think that since we already have that case of an exit with an incomplete bg_unlink, why not add another one, and possibly exit a bit faster in some scenario.

Reply:
There is a comment https://github.com/redis/redis/blob/unstable/src/server.c#L4915
We kill child process and persist rdb in main thread when we receive the signal at first time, we remove main thread dumping RDB at the second time.  So i think the second signal arriving is short from the first if user want to quit urgently, temp rdb will not be too big, right?

Yes, i also think it is a OS bug if OS can't release all things. In `prepareForShutdown`, we exactly may have a pending non-closed fd when exits, but i think main thread also need to do somethings such as dumping a new rdb, the fd will be closed by bio thread in most cases. Your suggestion also is ok theoretically, but i just feel it is not nice.

Reply:
And just as you said, we shouldn't block server to quit because of waiting file closed in `prepareForShutdown`, I think it make much sense. Compared with some program practice, keeping it exit safely is more important.

Reply:
let's see what @yossigo thinks and decide.

Reply:
I am less concerned with the case that involves `redis-server` running interactively and killed with `SIGINT`, because the worst thing that happens is that users wait a bit.

I am more more concerned about cases involving Redis running as a service and being terminated or restarted. Blocking for a long time in this case may have bigger undesired effects (potentially due to bugs or false assumptions elsewhere).

The thing is, at this point I'm not sure leaving the file open when exiting really solves this issue because as I mentioned earlier it's possible for the process to remain defunct until the filesystem operation is completed, so from the POV of anyone waiting on it could be the same. And worse, this could be OS/kernel version/filesystem dependent.

So I'm OK either way, and suggest to be aware and on the watch about this.

Reply:
So maybe we should add a comment in prepareShutdown that mentions that we exit without waiting for a possibly pending `bg_unlink`?

And if we do that, i'd suggest to do the same in SIGINT, mainly for consistency.
So i would vote for the patch i posted. arguing that it's more explicit and consistent with what happens in the other flow.

In theory this can smoke out problem if there are any (since that behavior will be more common)

Reply:
Everyone has their own programming taste, in my team, we also have this problem. For me, I‘m willing to respect the decision of the code repo owner after enough discussions since the owner finally is responsible for code repo. If not, things won't move forward.

So, if you (coreteam) reach an agreement, I will do it, after all, these solutions are not wrong in our perception.

Reply:
@ShooterIT we don't need core team consensus for this, it's enough that one of us approves it.
We have Yossi saying that he's ok with both options, and i prefer one because of what i listed in my last post.
So you can proceed and implement it if you ok with that decision.

Please don't give in to our judgement if you feel they're wrong, there's always a chance that we're missing something.
We appreciate your opinion and the hard work you're putting in improving redis!

Reply:
@oranagra  OK, let me do next job. There is no error on my opinion, your patch can let problem appear ASAP if has.
 
I have benefited a lot from Redis, I should also contribute to it :)